File: src/main/java/me/aov/sellgui/managers/RandomPriceManager.java
package me.aov.sellgui.managers;

import me.aov.sellgui.SellGUIMain;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.inventory.ItemStack;

import java.util.Random;

public class RandomPriceManager {

    private final SellGUIMain plugin;
    private final Random random;

    public RandomPriceManager(SellGUIMain plugin) {
        this.plugin = plugin;
        this.random = new Random();
    }

    public void setRandomPriceRange(ItemStack item, double minPrice, double maxPrice, String playerName) {
        if (item == null) return;

        String itemKey = getItemKey(item);
        String itemType = getItemType(item);

        FileConfiguration config = getRandomPricesConfig();
        if (config == null) return;

        String path = "prices." + itemKey;
        config.set(path + ".min_price", minPrice);
        config.set(path + ".max_price", maxPrice);
        config.set(path + ".last_updated", System.currentTimeMillis());
        config.set(path + ".set_by", playerName);
        config.set(path + ".item_type", itemType);

        saveConfig();

        plugin.getLogger().info("Set random price range for " + itemKey + ": $" + minPrice + " - $" + maxPrice);
    }

    public double getRandomPrice(ItemStack item) {
        if (item == null) return 0.0;

        String itemKey = getItemKey(item);
        FileConfiguration config = getRandomPricesConfig();
        if (config == null) return 0.0;

        String path = "prices." + itemKey;
        if (!config.contains(path)) {
            return 0.0;
        }

        double minPrice = config.getDouble(path + ".min_price", 0.0);
        double maxPrice = config.getDouble(path + ".max_price", 0.0);

        if (minPrice >= maxPrice) {
            return minPrice;
        }

        return minPrice + (random.nextDouble() * (maxPrice - minPrice));
    }

    public double[] getPriceRange(ItemStack item) {
        if (item == null) return null;

        String itemKey = getItemKey(item);
        FileConfiguration config = getRandomPricesConfig();
        if (config == null) return null;

        String path = "prices." + itemKey;
        if (!config.contains(path)) {
            return null;
        }

        double minPrice = config.getDouble(path + ".min_price", 0.0);
        double maxPrice = config.getDouble(path + ".max_price", 0.0);

        return new double[]{minPrice, maxPrice};
    }

    public boolean hasRandomPrice(ItemStack item) {
        if (item == null) return false;

        String itemKey = getItemKey(item);
        FileConfiguration config = getRandomPricesConfig();
        if (config == null) return false;

        return config.contains("prices." + itemKey);
    }

    public void removeRandomPrice(ItemStack item) {
        if (item == null) return;

        String itemKey = getItemKey(item);
        FileConfiguration config = getRandomPricesConfig();
        if (config == null) return;

        config.set("prices." + itemKey, null);
        saveConfig();

        plugin.getLogger().info("Removed random price range for " + itemKey);
    }

    private String getItemKey(ItemStack item) {

        if (hasMMOItems()) {
            try {
                String mmoId = getMMOItemsId(item);
                if (mmoId != null) {
                    return "mmoitems:" + mmoId;
                }
            } catch (Exception e) {

            }
        }

        if (hasNexo()) {
            try {
                String nexoId = getNexoId(item);
                if (nexoId != null) {
                    return "nexo:" + nexoId;
                }
            } catch (Exception e) {

            }
        }

        return item.getType().name();
    }

    private String getItemType(ItemStack item) {
        if (hasMMOItems()) {
            try {
                String mmoId = getMMOItemsId(item);
                if (mmoId != null) {
                    return "MMOITEMS";
                }
            } catch (Exception e) {

            }
        }

        if (hasNexo()) {
            try {
                String nexoId = getNexoId(item);
                if (nexoId != null) {
                    return "NEXO";
                }
            } catch (Exception e) {

            }
        }

        return "VANILLA";
    }

    private boolean hasMMOItems() {
        try {
            return plugin.getServer().getPluginManager().getPlugin("MMOItems") != null;
        } catch (Exception e) {
            return false;
        }
    }

    private boolean hasNexo() {
        try {
            return plugin.getServer().getPluginManager().getPlugin("Nexo") != null;
        } catch (Exception e) {
            return false;
        }
    }

    private String getMMOItemsId(ItemStack item) {
        try {

            if (item.hasItemMeta() && item.getItemMeta().hasDisplayName()) {
                String displayName = item.getItemMeta().getDisplayName();

                return null;
            }
            return null;
        } catch (Exception e) {
            return null;
        }
    }

    private String getNexoId(ItemStack item) {
        try {

            try {
                Class<?> nexoItemsClass = Class.forName("com.nexomc.nexo.api.NexoItems");
                java.lang.reflect.Method getIdMethod = nexoItemsClass.getMethod("idFromItem", ItemStack.class);
                Object result = getIdMethod.invoke(null, item);
                if (result != null) {
                    return result.toString();
                }
            } catch (Exception e) {

            }

            return null;
        } catch (Exception e) {
            plugin.getLogger().warning("Error detecting Nexo item: " + e.getMessage());
            return null;
        }
    }

    private FileConfiguration getRandomPricesConfig() {
        if (plugin.getConfigManager() != null) {
            return plugin.getConfigManager().getConfig("random-prices");
        }
        return null;
    }

    private void saveConfig() {
        if (plugin.getConfigManager() != null) {
            plugin.getConfigManager().saveConfig("random-prices");
        }
    }

    public void cleanupOldEntries() {
        FileConfiguration config = getRandomPricesConfig();
        if (config == null) return;

        boolean autoCleanup = config.getBoolean("config.auto_cleanup", true);
        if (!autoCleanup) return;

        int cleanupDays = config.getInt("config.cleanup_days", 30);
        long cutoffTime = System.currentTimeMillis() - (cleanupDays * 24 * 60 * 60 * 1000L);

        if (config.contains("prices")) {
            for (String key : config.getConfigurationSection("prices").getKeys(false)) {
                long lastUpdated = config.getLong("prices." + key + ".last_updated", 0);
                if (lastUpdated < cutoffTime) {
                    config.set("prices." + key, null);
                    plugin.getLogger().info("Cleaned up old random price entry: " + key);
                }
            }
        }

        saveConfig();
    }

    public int getStoredPricesCount() {
        FileConfiguration config = getRandomPricesConfig();
        if (config == null || !config.contains("prices")) return 0;

        return config.getConfigurationSection("prices").getKeys(false).size();
    }

    public boolean requiresEvaluation(ItemStack item) {
        return hasRandomPrice(item);
    }

    public boolean isEvaluated(ItemStack item) {
        if (item == null || !item.hasItemMeta()) return false;

        if (item.getItemMeta().getPersistentDataContainer() != null) {
            try {
                org.bukkit.NamespacedKey worthKey = new org.bukkit.NamespacedKey(plugin, "worth");
                if (item.getItemMeta().getPersistentDataContainer().has(worthKey,
                        org.bukkit.persistence.PersistentDataType.DOUBLE)) {
                    return true;
                }
            } catch (Exception e) {

            }
        }

        if (item.hasItemMeta() && item.getItemMeta().getPersistentDataContainer() != null) {
            try {
                org.bukkit.NamespacedKey evaluatedKey = new org.bukkit.NamespacedKey(plugin, "evaluated");
                if (item.getItemMeta().getPersistentDataContainer().has(evaluatedKey,
                        org.bukkit.persistence.PersistentDataType.BYTE)) {
                    return true;
                }
            } catch (Exception e) {

            }
        }

        try {
            if (plugin.getNBTPriceManager() != null) {
                double nbtPrice = plugin.getNBTPriceManager().getPriceFromNBT(item);
                return nbtPrice > 0;
            }
        } catch (Exception e) {

            plugin.getLogger().warning("Error checking NBT price: " + e.getMessage());
        }

        return false;
    }

    public boolean canBeSold(ItemStack item) {
        if (!requiresEvaluation(item)) {
            return true;
        }

        return isEvaluated(item);
    }
}

File: src/main/kotlin/me/aov/sellgui/managers/RandomPriceManager.kt
package me.aov.sellgui.managers

import me.aov.sellgui.SellGUIMain
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import java.util.Random

class RandomPriceManager(private val plugin: SellGUIMain) {
    private val random = Random()
    fun setRandomPriceRange(item: ItemStack?, minPrice: Double, maxPrice: Double, playerName: String) {
        item ?: return; val itemKey = getItemKey(item); val config = randomPricesConfig ?: return
        config.set("prices.$itemKey.min_price", minPrice); config.set("prices.$itemKey.max_price", maxPrice); config.set("prices.$itemKey.last_updated", System.currentTimeMillis()); config.set("prices.$itemKey.set_by", playerName); config.set("prices.$itemKey.item_type", getItemType(item)); saveConfig()
        plugin.logger.info("Set random price range for $itemKey: $$minPrice - $$maxPrice")
    }
    fun getRandomPrice(item: ItemStack?): Double { item ?: return 0.0; val config = randomPricesConfig ?: return 0.0; val path = "prices.${getItemKey(item)}"; if (!config.contains(path)) return 0.0; val min = config.getDouble("$path.min_price", 0.0); val max = config.getDouble("$path.max_price", 0.0); return if (min >= max) min else min + random.nextDouble() * (max - min) }
    fun getPriceRange(item: ItemStack?): DoubleArray? { item ?: return null; val config = randomPricesConfig ?: return null; val path = "prices.${getItemKey(item)}"; return if (config.contains(path)) doubleArrayOf(config.getDouble("$path.min_price", 0.0), config.getDouble("$path.max_price", 0.0)) else null }
    fun hasRandomPrice(item: ItemStack?): Boolean = item != null && randomPricesConfig?.contains("prices.${getItemKey(item)}") == true
    fun removeRandomPrice(item: ItemStack?) { item ?: return; randomPricesConfig?.set("prices.${getItemKey(item)}", null); saveConfig(); plugin.logger.info("Removed random price range for ${getItemKey(item)}") }
    private fun getItemKey(item: ItemStack): String = if (hasNexo()) getNexoId(item)?.let { "nexo:$it" } ?: item.type.name else item.type.name
    private fun getItemType(item: ItemStack): String = if (hasNexo() && getNexoId(item) != null) "NEXO" else "VANILLA"
    private fun hasNexo(): Boolean = try { plugin.server.pluginManager.getPlugin("Nexo") != null } catch (_: Exception) { false }
    private fun getNexoId(item: ItemStack): String? = try { Class.forName("com.nexomc.nexo.api.NexoItems").getMethod("idFromItem", ItemStack::class.java).invoke(null, item)?.toString() } catch (_: Exception) { null }
    private val randomPricesConfig get() = plugin.configManager?.getConfig("random-prices")
    private fun saveConfig() { plugin.configManager?.saveConfig("random-prices") }
    fun cleanupOldEntries() { val config = randomPricesConfig ?: return; if (!config.getBoolean("config.auto_cleanup", true)) return; val cutoff = System.currentTimeMillis() - config.getInt("config.cleanup_days", 30) * 86_400_000L; config.getConfigurationSection("prices")?.getKeys(false)?.forEach { key -> if (config.getLong("prices.$key.last_updated", 0) < cutoff) { config.set("prices.$key", null); plugin.logger.info("Cleaned up old random price entry: $key") } }; saveConfig() }
    fun getStoredPricesCount(): Int = randomPricesConfig?.getConfigurationSection("prices")?.getKeys(false)?.size ?: 0
    fun requiresEvaluation(item: ItemStack?): Boolean = hasRandomPrice(item)
    fun isEvaluated(item: ItemStack?): Boolean {
        if (item == null || !item.hasItemMeta()) return false
        val container = item.itemMeta.persistentDataContainer
        val hasPrice = container.has(NamespacedKey(plugin, "worth"), PersistentDataType.DOUBLE) || (container.get(NamespacedKey(plugin, "current_price"), PersistentDataType.DOUBLE) ?: 0.0) > 0 || container.has(NamespacedKey(plugin, "evaluated"), PersistentDataType.BYTE)
        return hasPrice || (plugin.getNBTPriceManager()?.getPriceFromNBT(item) ?: 0.0) > 0
    }
    fun canBeSold(item: ItemStack?): Boolean = !requiresEvaluation(item) || isEvaluated(item)
}
